# Tutorial: How to set up a geospatial collection of 3D Hydrography Program data with ckanext-gztr using Python, Apache SedonaDB

In this tutorial we will go over how to download 3D Hydrography Program (3DHP) data for your CKAN instance's region.

We expect you to download the 3DHP Continental US ZIP file published for the Gregorian year 2026 which is about 20GB, and when you uncompress it there should be a GeoPackage (`.gpkg`) file which is about 60GB.

You can download the ZIP file from the USGS ScienceBase catalog [here](https://www.sciencebase.gov/catalog/item/69743a65d4be0260181a121e) under the `Related External Resources` section where there is a link under `Type: Download`.

Learn more:

- ckanext-gztr: [gztr.dathere.com](https://gztr.dathere.com)
- 3DHP: [usgs.gov/3dhp](https://www.usgs.gov/3dhp)

## Setup your libraries and SedonaDB connection

First we'll install external Python libraries:

- [Apache SedonaDB](https://sedona.apache.org/sedonadb/latest/) - For setting up a geospatial [DataFrame](https://sedona.apache.org/sedonadb/latest/reference/python/#sedonadb.dataframe.DataFrame)
- [lonboard](https://developmentseed.org/lonboard/latest) - For visualizing our DataFrame in an interactive map
- [jupyterlab](https://pypi.org/project/jupyterlab/) - Required for lonboard to work properly in a Jupyter Lab (based on [this issue](https://github.com/developmentseed/lonboard/issues/397))
- [ckanapi](https://github.com/ckan/ckanapi) - For interacting with your CKAN instance's API (to upload the data and modify your STAC Catalog)
- [requests](https://requests.readthedocs.io/en/latest/) - For retrieving your STAC `catalog.json` and `collections.json` data when we update them

In [1]:
# The dependencies should already be installed if you're using mybinder.org
#   If not then uncomment and run one of the following commands.
#   Then you'll need to refresh your browser to use the interactive lonboard visualizations.
# !pip install "apache-sedona[db]" lonboard jupyterlab ckanapi requests
# or if you have uv installed
# !uv pip install "apache-sedona[db]" lonboard jupyterlab ckanapi requests

We'll import the libraries.

In [2]:
import requests
import sedona.db
from ckanapi import RemoteCKAN
from lonboard import viz
from pprint import pprint

We'll initialize our SedonaDB connection.

In [3]:
sd = sedona.db.connect()

Now we'll read the GeoPackage file that we've extracted. You can choose from several layers and pass them to the `layer` option in the `read_pyogrio` function including:

- `hydro_3dhp_all_flowline`
- `hydro_3dhp_all_rc_team`
- `hydro_3dhp_all_waterbody`
- `hydro_3dhp_all_landscape`
- `hydro_3dhp_all_network`
- `hydro_3dhp_all_catchment`
- `hydro_3dhp_all_drainagearea`

For our tutorial we'll use `hydro_3dhp_all_waterbody`.

In [4]:
df = sd.read_pyogrio("./3dhp_all_CONUS_20260112_GPKG/3dhp_all_CONUS_20260112_GPKG.gpkg", options={"layer":"hydro_3dhp_all_waterbody"})

Let's explore the columns that are available.

In [5]:
df.schema

SedonaSchema with 12 fields:
  id3dhp: utf8<Utf8>
  featuredate: timestamp<Timestamp(ms, "UTC")>
  mainstemid: utf8<Utf8>
  gnisid: int32<Int32>
  gnisidlabel: utf8<Utf8>
  featuretype: int16<Int16>
  featuretypelabel: utf8<Utf8>
  areasqkm: float64<Float64>
  workunitid: utf8<Utf8>
  shape_Length: float64<Float64>
  shape_Area: float64<Float64>
  shape: geometry<Wkb(epsg:6350)>

Notice that the only column with a `geometry` type is the `shape` column, and its Coordinate Reference System (CRS) is `epsg:6350`. Currently for ckanext-gztr the required CRS is `epsg:4326`, so we'll perform a few steps to transform the CRS and also remove a few unnecessary columns.

First we'll define a new view named `data` that we can refer to when running spatial SQL queries for our data.

In [6]:
df.to_view("data", overwrite=True)

Now we'll define a variable `selected_columns` to get columns that do not start with `shape`.

In [7]:
selected_columns = [col for col in df.columns if not col.startswith("shape")]

Now we'll select all non-shape columns and use [ST_Transform](https://sedona.apache.org/sedonadb/latest/reference/sql/#st_transform) to transform the CRS to the intended one. We'll also overwrite our `data` view with the new data.

In [8]:
df = sd.sql(f"""
SELECT {",".join(selected_columns)}, ST_TRANSFORM("shape",'EPSG:4326') as "shape" FROM data
""")
df.to_view("data", overwrite=True)

Just to understand how many 3DHP features are in our data, we can run a count.

In [9]:
sd.sql("""
SELECT COUNT("id3dhp") FROM data
""").show()

┌────────────────────┐
│ count(data.id3dhp) │
│        int64       │
╞════════════════════╡
│            5747966 │
└────────────────────┘


## Retrieve your region's bounding box from [geojson.io](https://geojson.io)

Our CKAN instance is (likely) not globally-scoped, but rather it exists for storing a particular region's water data. Therefore we want to filter for the CKAN instance's region of interest and download a subset of the 3DHP data instead of all of the available data.

For retrieving 3DHP data that intersects with our region, we'll need to get bounding box coordinate values. In the video below we provide an example using [geojson.io](https://geojson.io) to draw a bounding box over our region of interest and copy its bounding box value.

**IMPORTANT**: For the last step instead of copying the bounding box `BBOX`, **copy the `WKT` value**.

In [10]:
from IPython.display import Video

Video("../media/geojson_io_bounding_box_demo_lq.mp4")

In [11]:
# Filter on New Mexico bounding box
# Replace the value of bounding_box_wkt with your bounding box
bounding_box_wkt = "POLYGON ((-109.08617242628395 37.02865746972863,-103.00865705250497 37.028715659351136,-103.00817441210035 31.292013555518253,-109.08616254688481 31.29254469262679,-109.08617242628395 37.02865746972863))"
df = sd.sql(f"""
SELECT * FROM data
WHERE ST_Intersects("shape",ST_GeomFromWKT('{bounding_box_wkt}', 'EPSG:4326')) = true
""")
df.to_view("data", overwrite=True)

In [12]:
sd.sql("""
SELECT COUNT(*) FROM data
""").show()

┌──────────┐
│ count(*) │
│   int64  │
╞══════════╡
│    78238 │
└──────────┘


Let's show an interactive visualization of our dataframe using lonboard. You can click on a feature on the map to view its properties.

In [13]:
viz(df.to_pandas())

In [14]:
df.count()

78238

Some features may not intersect with our region (e.g. the bounding box covers areas that are not within New Mexico's boundaries). You can look into the [`df.filter`](https://sedona.apache.org/sedonadb/latest/reference/python/#sedonadb.dataframe.DataFrame.filter) function to assist with this. For example to filter out the feature with the `id3dhp` ID of `J2GN4`, you can run `df = df.filter(df.id3dhp != "J2GN4")`.

To format the data properly for compliance with ckanext-gztr and our STAC specification setup, we'll need to:

- Rename the `id3dhp` column to `id`
- Rename the `gnisidlabel` column to `title`
- Rename the `mainstemid` column to `geoconnex_uri`
- Rename the `shape` column to `geometry`
- Temporarily we'll keep `featuredate` for calculating the temporal extent of our filtered data

To perform this transformation, we'll:

- Reassign our dataframe `df` to the output of a spatial SQL query that fulfills the aforementioned renaming requirements

In [15]:
df = sd.sql("""
SELECT "id3dhp" as "id",featuredate,"gnisidlabel" as "title","mainstemid" AS "geoconnex_uri","shape" AS "geometry" FROM data
""")
df.to_view("data", overwrite=True)

In [16]:
df.schema

SedonaSchema with 5 fields:
  id: utf8<Utf8>
  featuredate: timestamp<Timestamp(ms, "UTC")>
  title: utf8<Utf8>
  geoconnex_uri: utf8<Utf8>
  geometry: geometry<Wkb(ogc:crs84)>

In [17]:
df.count()

78238

There are many features that do not have a `gnisidlabel` value. For example here we run a count of how many features have a `gnisidlabel` value.

In [18]:
sd.sql("""
SELECT * FROM data
WHERE "title" != ''
""").count()

6210

If you were to filter further then even less have both a `gnisidlabel` and a Geoconnex URI.

In [19]:
sd.sql("""
SELECT * FROM data
WHERE "title" != '' AND "geoconnex_uri" != ''
""").count()

4

In our case we don't need to worry about missing Geoconnex URI values for loading the 3DHP data into ckanext-gztr. However we do need a `title` value for each geospatial feature. Therefore we can replace the `title` value for each 3DHP feature with the `id3dhp` value, which in our view is the `id` field. The existing `title` values we can keep with the original `gnisidlabel` values (or replace them too if you'd like). You could also opt to only keep features that have a `gnisidlabel` value.

For lower file size and demo purposes we will keep only 3DHP features that have a `title` value.

In [20]:
df = sd.sql("""
SELECT * FROM data
WHERE "title" != ''
""")
df.to_view("data", overwrite=True)
df.show()

┌───────┬──────────────────────┬───────────────┬───────────────┬───────────────────────────────────┐
│   id  ┆      featuredate     ┆     title     ┆ geoconnex_uri ┆              geometry             │
│  utf8 ┆       timestamp      ┆      utf8     ┆      utf8     ┆              geometry             │
╞═══════╪══════════════════════╪═══════════════╪═══════════════╪═══════════════════════════════════╡
│ IBBQE ┆ 2023-09-14T00:00:00Z ┆ Bobby Lake    ┆               ┆ MULTIPOLYGON Z(((-103.4556613013… │
├╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ IBD5O ┆ 2023-09-14T00:00:00Z ┆ Slash F Tank  ┆               ┆ MULTIPOLYGON Z(((-103.2135669027… │
├╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ IE46L ┆ 2023-09-14T00:00:00Z ┆ Morgan Lake   ┆               ┆ MULTIPOLYGON Z(((-104.2142974997… │
├╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌

In [21]:
df.count()

6210

We now have the data formatted as required for usage with ckanext-gztr.

Before we continue we'll want to get the start date and end date for the temporal extent when we set up the STAC collection for this geospatial collection.

In [22]:
import datetime

# Get a dataframe of all distinct timestamps and sort them from earliest to latest as a GeoPandas dataframe (so we can use the .iloc method)
distinct_timestamps_df = sd.sql("SELECT featuredate FROM data").distinct_on("featuredate").sort("featuredate").to_pandas()

# Get the first and the last timestamp value from the sorted dataframe
start_time = datetime.datetime.strptime(str(distinct_timestamps_df.iloc[0]["featuredate"]), "%Y-%m-%d %H:%M:%S%z").strftime("%Y-%m-%dT%H:%M:%SZ")
end_time = datetime.datetime.strptime(str(distinct_timestamps_df.iloc[-1]["featuredate"]), "%Y-%m-%d %H:%M:%S%z").strftime("%Y-%m-%dT%H:%M:%SZ")

print(f"Start time: {start_time}")
print(f"End time: {end_time}")

Start time: 2023-09-14T00:00:00Z
End time: 2023-09-14T00:00:00Z


Now we can optionally remove the `featuredate` column.

In [24]:
df = sd.sql("""
SELECT id,title,geoconnex_uri,geometry FROM data
WHERE "title" != ''
""")
df.to_view("data", overwrite=True)
df.schema

SedonaSchema with 4 fields:
  id: utf8<Utf8>
  title: utf8<Utf8>
  geoconnex_uri: utf8<Utf8>
  geometry: geometry<Wkb(ogc:crs84)>

You can download this dataframe as a GeoParquet file with `df.to_parquet()`.

In [25]:
df.to_parquet("nm_3dhp_waterbodies.parquet")

## Upload your GeoParquet file to your CKAN instance and update your STAC Catalog (`catalog.json` and `collections.json`)

We have now:

- **Extract**ed our region's 3DHP waterbodies data from the downloaded 3DHP GeoPackage file
- **Transform**ed our data to include only 3DHP features that have a `gnisidlabel` value and renamed the columns to be compliant with ckanext-gztr and STAC

Now we need to perform the final 2 steps:

- **Load** the GeoParquet file `nm_3dhp_waterbodies.parquet` to our CKAN instance's STAC Catalog
- **Update** our STAC Catalog `catalog.json` file and `collections.json` file to expose our new geospatial collection's data to make the data accessible from ckanext-gztr and the STAC API

### Upload your GeoParquet file to your CKAN instance's storage

**IMPORTANT NOTE: We highly recommend to run the rest of the steps on a local device rather than through this online session of Jupyter Lab.** This is because you'll need to provide your CKAN API key (as shown in the cell below) to run the ckanapi commands. In the meantime you can read the rest of this notebook to learn how the load and update steps would work, and copy them into your own local Jupyter Lab or Python script if needed.

In [26]:
CKAN_INSTANCE_URL = "http://localhost:5000"
CKAN_API_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJqdGkiOiJjeXhpc1g2SHlOZjNkQzlPZWZXekZNbHE2Z0Nham5vS3ZTaktFTnZ3ZnNnIiwiaWF0IjoxNzg3MDUzMTcwfQ.EUftyGDkEx312TA_5S-FxcIaYDwIgiNSyASEi-juoVk"

ckan = RemoteCKAN(CKAN_INSTANCE_URL, apikey=CKAN_API_KEY)

In [27]:
file_create_response = ckan.action.file_create(storage="gztr", upload=open("nm_3dhp_waterbodies.parquet", "rb"))
pprint(file_create_response)

{'algorithm': 'md5',
 'content_type': 'application/octet-stream',
 'created': '2026-09-01T17:36:11.010640+00:00',
 'hash': '04bbe7fb1fcf4ec557a2aae4fb5a02ad',
 'id': 'edea81fd-e409-4a3f-88d1-a4e755252a56',
 'location': 'nm_3dhp_waterbodies.parquet',
 'name': 'nm_3dhp_waterbodies.parquet',
 'owner_id': '255d924c-f5fc-4dc1-9204-1e9bfa899219',
 'owner_type': 'user',
 'pinned': False,
 'size': 2611637,
 'storage': 'gztr',
 'storage_data': {}}


Store the location stem value as this will be important for updating our STAC files.

In [28]:
file_location_stem = file_create_response["location"].split(".")[0]
file_location_stem

'nm_3dhp_waterbodies'

In [29]:
# == Troubleshooting ==
# If you need to reupload your GeoParquet file then uncomment and run this cell
# you can use file_owner_scan to get the ID of the file and then file_delete to delete by file ID
# storage_files_to_check = ckan.action.file_owner_scan(storage="gztr")["results"]
# pprint(storage_files_to_check)
# ckan.action.file_delete(storage="gztr", id="")

### Update your STAC files (`catalog.json` and `collections.json`)

We'll need to update our STAC Catalog by updating the `catalog.json` and `collections.json` files stored in our CKAN instance's storage.

First we'll prepare `catalog` and `collections` variables and then we'll update our files.

First we'll retrieve the `catalog.json` file data as a Python object.

In [30]:
catalog = requests.get(f"{CKAN_INSTANCE_URL}/gztr/stac").json()

Now we'll want to append a new dictionary to our STAC Catalog's `links` list which points to our new HUC08 Sub Basins STAC collection.

You can refer to the STAC links specification for guidelines about each part of the dictionary here: https://github.com/radiantearth/stac-spec/blob/master/commons/links.md#link-object

In [31]:
COLLECTION_TITLE = "3DHP Waterbodies"

In [32]:
# Check if collection already exists in STAC Catalog
for link in catalog["links"]:
    if link["href"] == f"{CKAN_INSTANCE_URL}/gztr/stac/collections/{file_location_stem}":
        raise Exception("Collection already exists in catalog.json.")

catalog["links"].append({
    "href": f"{CKAN_INSTANCE_URL}/gztr/stac/collections/{file_location_stem}",
    "rel": "child",
    "title": COLLECTION_TITLE,
    "type": "application/json"
})
pprint(catalog)

{'description': 'Geospatial collections and features used for dataset '
                'publishing and search by the New Mexico Water Data Catalog, '
                'organized through the SpatioTemporal Asset Catalogs (STAC) '
                'specification.',
 'id': 'nmwdc',
 'links': [{'href': 'http://localhost:5000/gztr/stac',
            'rel': 'self',
            'type': 'application/json'},
           {'href': 'http://localhost:5000/gztr/stac',
            'rel': 'root',
            'type': 'application/json'},
           {'href': 'http://localhost:5000/gztr/stac/collections',
            'rel': 'collections',
            'type': 'application/json'},
           {'href': 'http://localhost:5000/gztr/stac/collections/public_water_systems',
            'rel': 'child',
            'title': 'Public Water Systems',
            'type': 'application/json'},
           {'href': 'http://localhost:5000/gztr/stac/collections/nm_huc08_sub_basins',
            'rel': 'child',
            'tit

Now we'll need to retrieve our `collections.json` file data.

In [33]:
collections = requests.get(f"{CKAN_INSTANCE_URL}/gztr/stac/collections").json()

The `collections.json` file's data is a list of dictionaries where each dictionary represents the metadata for a STAC collection. In this case we'll append a new dictionary for our HUC08 Sub Basins. Make sure to update the values in this dictionary as per your use case.

You can refer to the STAC Collections specification for guidelines about each part of the dictionary here: https://github.com/radiantearth/stac-api-spec/blob/v1.0.0/stac-spec/collection-spec/collection-spec.md

First we'll calculate the spatial and temporal extent that will be used in the STAC collection's `extent` value as [defined in the specification here](https://github.com/radiantearth/stac-spec/blob/master/collection-spec/collection-spec.md#extents).

We'll get the `bbox` value as a variable named `bbox`.

In [34]:
# Calculate the spatial extent (extent.spatial.bbox)
bbox = df.to_pandas().total_bounds.tolist()
bbox

[-109.10331128889611,
 31.29595836684165,
 -103.02752490282423,
 37.056743358427155]

For the temporal extent you can base it on the `featuredate` column's minimum and maximum value.

In [35]:
COLLECTION_DESCRIPTION = "3D Hydrography Program (3DHP) waterbodies intersecting with the New Mexico region. Filtered for only 3DHP features that have a Geographic Names Information System (GNIS) label."

collections.append({
  "stac_version": "1.1.0",
  "title": COLLECTION_TITLE,
  "type": "Collection",
  "description": COLLECTION_DESCRIPTION,
  "extent": {
    "spatial": {
      "bbox": [bbox]
    },
    "temporal": {
      "interval": [[start_time,None]]
    }
  },
  "id": file_location_stem,
  "license": "CC0-1.0",
  "links": [
    {
      "href": f"{CKAN_INSTANCE_URL}/gztr/stac/collections/{file_location_stem}",
      "rel": "self",
      "type": "application/json"
    },
    {
      "href": f"{CKAN_INSTANCE_URL}/gztr/stac/collections",
      "rel": "parent",
      "type": "application/json"
    },
    {
      "href": f"{CKAN_INSTANCE_URL}/gztr/stac",
      "rel": "root",
      "type": "application/json"
    },
    {
      "href": f"{CKAN_INSTANCE_URL}/gztr/stac/collections/{file_location_stem}/items",
      "rel": "items",
      "type": "application/geo+json"
    }
  ],
  "providers": [
    {
      "description": "An agency of the U.S. Department of the Interior. The USGS studies U.S. lands and resources and monitors, analyzes, and predicts Earth’s changing systems while providing data for science.",
      "name": "U.S. Geological Survey",
      "roles": [
        "producer",
        "host"
      ],
      "url": "https://usgs.gov"
    }
  ]
})

Now we'll need to write our `catalog` and `collections` variables to local JSON files before we upload them to our CKAN instance.

In [36]:
import json

with open("catalog.json", "w") as f:
    json.dump(catalog, f)
with open("collections.json", "w") as f:
    json.dump(collections, f)

Now we'll want to delete the existing `catalog.json` and `collections.json` files and replace them with our new files.

In [37]:
storage_files = ckan.action.file_owner_scan(storage="gztr")["results"]

In [38]:
old_catalog_file_id = [file for file in storage_files if file["location"] == "catalog.json"][0]["id"]
ckan.action.file_delete(storage="gztr", id=old_catalog_file_id)
new_catalog_create_response = ckan.action.file_create(storage="gztr", upload=open("catalog.json", "rb"))

In [39]:
old_collections_file_id = [file for file in storage_files if file["location"] == "collections.json"][0]["id"]
ckan.action.file_delete(storage="gztr", id=old_collections_file_id)
new_collections_create_response = ckan.action.file_create(storage="gztr", upload=open("collections.json", "rb"))

Now your geospatial collection should appear on your CKAN instance's gazetteer maps and through your STAC API!

In [40]:
new_catalog = requests.get(f"{CKAN_INSTANCE_URL}/gztr/stac").json()
new_catalog

{'description': 'Geospatial collections and features used for dataset publishing and search by the New Mexico Water Data Catalog, organized through the SpatioTemporal Asset Catalogs (STAC) specification.',
 'id': 'nmwdc',
 'links': [{'href': 'http://localhost:5000/gztr/stac',
   'rel': 'self',
   'type': 'application/json'},
  {'href': 'http://localhost:5000/gztr/stac',
   'rel': 'root',
   'type': 'application/json'},
  {'href': 'http://localhost:5000/gztr/stac/collections',
   'rel': 'collections',
   'type': 'application/json'},
  {'href': 'http://localhost:5000/gztr/stac/collections/public_water_systems',
   'rel': 'child',
   'title': 'Public Water Systems',
   'type': 'application/json'},
  {'href': 'http://localhost:5000/gztr/stac/collections/nm_huc08_sub_basins',
   'rel': 'child',
   'title': 'HUC08 Sub Basins',
   'type': 'application/json'},
  {'href': 'http://localhost:5000/gztr/stac/collections/ny_counties',
   'rel': 'child',
   'title': 'New York Counties',
   'type': 'a

## What next?

Explore the ckanext-gztr documentation at [gztr.dathere.com](https://gztr.dathere.com).